llvm: openmp/clang changes

DomT4 opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally prior to submission with `brew install <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [x] Does your submission pass `brew audit --strict --online <formula>` (after doing `brew install <formula>`)?

-----

Changes:
* Makes `openmp` a complete default.
* Makes `clang` a complete default.
* Fix some nits.
* Delete the `clang-omp` formula.

Need to do a complete rebuild locally but looks good.

CC @xu-cheng @MikeMcQuaid 

Closes https://github.com/Homebrew/homebrew-core/issues/1254.

MikeMcQuaid commented:
LGTM 👍

DomT4 commented:
I think `libiomp` is safe to be killed alongside `clang-omp`; LLVM seems to contain those elements now as well.

xu-cheng commented on Formula/llvm.rb:
```diff
@@ -284,16 +274,18 @@ def install
+
+    # Create symlinks for compatibility with prior "clang-omp" formula.
+    bin.install_symlink bin/"clang" => "clang-omp"
+    bin.install_symlink bin/"clang" => "clang-omp++"
```
I don't think this is necessary. Clang-omp makes it very clear that it is just a temporary project before openmp supporting lands in llvm. In short, llvm and clang-omp are two distinguish projects.

DomT4 replied:
I didn't know if I was worth preserving the links temporarily? But I guess since this formula is `keg_only` and `clang-omp` wasn't we wouldn't automatically be saving any scripts that had come to use that symlink. Happy to remove.

xu-cheng commented:
Just one comment :+1:

DomT4 commented:
```
==> Summary
🍺  /usr/local/Cellar/llvm/3.8.0: 1,815 files, 574.1M, built in 16 minutes 42 seconds

~> brew test -v llvm
$?
0
```
CI is happy as well. Any objections to shipping, feel free to shout, otherwise I'll merge later tonight.

maxim-belkin commented:
I believe you don't need `-liomp5`

maxim-belkin commented:
You might also want to add `-nobuiltininc`, because otherwise clang's builtin path `/usr/local/Cellar/llvm/3.8.0/bin/../lib/clang/3.8.0/include` is searched even without your `-I...` flag

maxim-belkin commented:
I forgot about a few changes I wanted to add in my PR. Please consider adding them here since you are going to update llvm to 3.8.1:
1. change `-DLLVM_ABI_BREAKING_CHECKS=WITH_ASSERTS`
2. add `-DLLVM_BUILD_EXAMPLES=ON` (why not?)
3. add `-DLLVM_ENABLE_EH=ON`
4. add `-DLLVM_CREATE_XCODE_TOOLCHAIN=ON` 

Their description is [here](http://llvm.org/releases/3.8.1/docs/CMake.html)

DomT4 commented:
> change -DLLVM_ABI_BREAKING_CHECKS=WITH_ASSERTS

Can you elaborate on whether this is already the default or not? The line:

> A version of LLVM built with ABI breaking checks is not ABI compatible with a version built without it.

Doesn't give me a warm, fuzzy feeling if they aren't.

> 2. add -DLLVM_BUILD_EXAMPLES=ON (why not?)

Possibly space saving, depending on how large the examples are. Will check locally.

> add -DLLVM_CREATE_XCODE_TOOLCHAIN=ON

> This target will create a directory at $CMAKE_INSTALL_PREFIX/Toolchains containing an xctoolchain directory which can be used to override the default system tools.

Can you elaborate on how safe this is?

DomT4 commented:
Heh. Examples don't build:
```
/tmp/llvm-20160716-76629-1sbnz74/llvm-3.8.1.src/examples/Kaleidoscope/Chapter2/toy.cpp:68:14: error: use of undeclared identifier 'strtod'; did you mean 'strtok'?
    NumVal = strtod(NumStr.c_str(), nullptr);
             ^~~~~~
             strtok
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/usr/include/string.h:90:7: note: 'strtok' declared here
char    *strtok(char *, const char *);
         ^
/tmp/llvm-20160716-76629-1sbnz74/llvm-3.8.1.src/examples/Kaleidoscope/Chapter2/toy.cpp:68:21: error: cannot initialize a parameter of type 'char *' with an rvalue of type 'const value_type *' (aka 'const char *')
    NumVal = strtod(NumStr.c_str(), nullptr);
                    ^~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/usr/include/string.h:90:20: note: passing argument to parameter here
char    *strtok(char *, const char *);
                      ^
2 errors generated.
make[2]: *** [examples/Kaleidoscope/Chapter2/CMakeFiles/Kaleidoscope-Ch2.dir/toy.cpp.o] Error 1
make[1]: *** [examples/Kaleidoscope/Chapter2/CMakeFiles/Kaleidoscope-Ch2.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

maxim-belkin commented:
`-DLLVM_ABI_BREAKING_CHECKS=WITH_ASSERTS` is actually the default but our `if build.with? "test"` section changes it to `...=ON`, which is ... not correct: it can be `WITH_ASSERTS`, `FORCE_ON`, or `FORCE_OFF`

Additional toolchain folder -- seem safe to me...
I'll check the problem with examples.

ilovezfs commented:
>-DLLVM_CREATE_XCODE_TOOLCHAIN=ON

and the plan with that is ...

xu-cheng commented:
Also should clang-omp and libiomp be moved to boneyard instead of just being deleted.

ilovezfs commented:
Why would boneyarding them be useful?

ilovezfs commented:
Since openmp default is ON, what's the use case?

maxim-belkin commented:
I have successfully built examples. Here are my [gist-logs](https://gist.github.com/d6b9c9d1e2b4889a54af64fbf1dd8988).
I also set the parameter to build xcode toolchain but I don't see that folder. The reasoning behind building it is explained at the top of [this file](https://llvm.org/svn/llvm-project/llvm/trunk/tools/xcode-toolchain/CMakeLists.txt)

maxim-belkin commented:
I guess, I should have added `system "make", "install-xcode-toolchain"` after `system "make", "install"`

maxim-belkin commented:
Yes, `system "make", "install-xcode-toolchain"` installs toolchain files in `/usr/local/Cellar/llvm/3.8.0/Toolchains/`

This "Toolchain" directory has 118 subdirectories, 733 MB.

maxim-belkin commented:
how about `with-toolchain` option?

DomT4 commented:
Ooft. Yeah, that's definitely option territory.

Can you elaborate on how you got examples building?

maxim-belkin commented:
I didn't do anything special to build examples. But I built them for llvm 3.8.0, not 3.8.1.

maxim-belkin commented:
I'm testing this code.

maxim-belkin commented:
Got the same error as you. Will look into why it fails later - might be an upstream bug.

DomT4 commented:
* Moved `clang-omp` and `libiomp` to boneyard rather than full deletion.
* Added `with-toolchain` option.
* Revision bumped `creduce` because that seems to fix the CI test failure.
* Rebased & repushed.

DomT4 commented:
Happy to accept a PR to add the examples, but going to merge this now before someone else files a PR or something else finds a magical way to break. A green LLVM CI build is something to be cherished 😆.
